Spotbugs: NP_NONNULL_PARAM_VIOLATION

Fokko opened the pull request:
Enables `NP_NONNULL_PARAM_VIOLATION` in Spotbugs

Fokko commented:
Works on my CI: https://travis-ci.org/Fokko/druid/builds/562448511

Fokko commented:
Fixed conflicts

leventov commented on processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java:
```diff
@@ -450,18 +450,17 @@ public Row apply(final Row row)
             }, subtotalQuery, null),
-            subtotalQuery
-                             )
+            subtotalQuery)
```
This method call is formatted improperly. It should be
```java
results = mergeResults(
    new QueryRunner<Row>()
    {
        @Override
        public Sequence<Row> run(QueryPlus<Row> queryPlus, ResponseContext responseContext)
        {
          return GroupByRowProcessor.getRowsFromGrouper(query, null, grouperSupplier);
        }
    },
    subtotalQuery,
    null
);
subtotalsResults.add(applyPostProcessing(results, subtotalQuery));
```

Fokko replied:
Good catch, thanks

leventov commented on processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java:
```diff
@@ -450,18 +450,17 @@ public Row apply(final Row row)
       return Sequences.withBaggage(
           Sequences.concat(subtotalsResults),
-          () -> Lists.reverse(closeOnExit).forEach(closeable -> CloseQuietly.close(closeable))
+          () -> Lists.reverse(closeOnExit).forEach(CloseQuietly::close)
```
It's better to use Closer to close multiple closeables

leventov commented on processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java:
```diff
@@ -450,18 +450,17 @@ public Row apply(final Row row)
     }
     catch (Exception ex) {
-      Lists.reverse(closeOnExit).forEach(closeable -> CloseQuietly.close(closeable));
+      Lists.reverse(closeOnExit).forEach(CloseQuietly::close);
```
It's better to use Closer to close multiple closeables

Fokko commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchEstimatePostAggregator.java:
```diff
@@ -67,17 +70,10 @@ public Comparator getComparator()
-          );
-        }
-      };
+      return (o1, o2) -> Doubles.compare(
```
I don't see how to implement `Comparator.comparingDouble` here easily. I'll just revert to the original code. I think lambdas are much easier to read, therefore I've changed it.

Fokko commented on extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/ApproximateHistogram.java:
```diff
@@ -532,9 +534,9 @@ public ApproximateHistogram copy(ApproximateHistogram h)
-      float[] mergedPositions,
-      long[] mergedBins,
-      float[] deltas
+      @Nullable float[] mergedPositions,
```
It is because this constructor:
https://github.com/apache/incubator-druid/blob/782863ed0f00fe199599b6bafc587afd929e0f14/extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/ApproximateHistogram.java#L475

Which makes this one `@Nullable`:
https://github.com/apache/incubator-druid/blob/782863ed0f00fe199599b6bafc587afd929e0f14/extensions-core/histogram/src/main/java/org/apache/druid/query/aggregation/histogram/ApproximateHistogram.java#L478-L485

Which trickles down to making `mergedPositions` nullable.

Fokko replied:
This constructor is only used in tests and benchmarks.

Fokko replied:
We can drop zero-sized arrays instead and drop in the `@Visible for testing`. Which is nicer in my opinion. Would love to get your opinion on this.

Fokko replied:
Fixed this by removing the constructor entirely.

leventov replied:
Thanks

Fokko commented on processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java:
```diff
@@ -377,7 +377,7 @@ public Row apply(final Row row)
     }
     catch (Exception ex) {
-      Lists.reverse(closeOnExit).forEach(closeable -> CloseQuietly.close(closeable));
+      Lists.reverse(closeOnExit).forEach(CloseQuietly::close);
```
The code did not change, but we're the Java8 reference function notation: https://www.codementor.io/eh3rrera/using-java-8-method-reference-du10866vx

leventov replied:
I've created #8226 about this.

Fokko commented:
Rebased as well

Fokko commented:
@leventov I've removed all the unrelated changes and rebased against latest master. Let me know what you think. Would be great to get this in so I can continue on fixing all the `[Not]Null` annotations.

Fokko commented:
I guess a flakey test, it passed on my branch: https://travis-ci.org/Fokko/druid/builds/566424524

asdf2014 commented:
@Fokko I just restarted the failed travis job.
